Add cut evolution

bpalmeiro opened the pull request:
This PR will add efficiencies for S1, S2 and band cuts to the time evolution table. It is meant to be backwards compatible so the old configs work.

bpalmeiro committed:
Cosmetics and remove unused import

bpalmeiro committed:
Add a default value for normalization when creating histograms

bpalmeiro committed:
Add DataClass to store masks

bpalmeiro committed:
Add masks return when cutting

bpalmeiro committed:
Add function to compute eff time evolution

bpalmeiro committed:
Return new data frame

bpalmeiro commented:
I wonder if there is any test checking time evolution table contents, but I found none, any help here @ausonandres?

ausonandres commented:
No, there isn't.

bpalmeiro commented:
adding one, then

bpalmeiro committed:
Add test checking time evolution table content

bpalmeiro committed:
Add test checking efficiency is <=1

This test checks that cut efficiencies are less than 1, as any eff
should be.

mmkekic commented on krcal/core/kr_parevol_functions.py:
```diff
@@ -215,3 +217,25 @@ def kr_time_evolution(ts         : np.array,
 
     return total_pars
+
+def cut_time_evolution(masks_time : List[np.array],
```
Maybe we should try to use masks names consistently

mmkekic commented on krcal/core/kr_parevol_functions.py:
```diff
@@ -215,3 +217,25 @@ def kr_time_evolution(ts         : np.array,
+                       masks_cuts : masks_container,
+                       pars_table : pd.DataFrame):
+
+    len_ts = len(masks_time)
```
align

mmkekic commented on krcal/map_builder/map_builder_functions.py:
```diff
@@ -497,12 +493,13 @@ def remove_peripheral(map       : ASectorMap,
 
 def add_krevol(maps         : ASectorMap,
                dst          : pd.DataFrame,
+               masks_cuts   : masks_container,
```
this parameter is not in docstrings, also why we dont have docstrings for all the functions in this file?

bpalmeiro replied:
It's a good question, maybe a new PR adding documentation. I did also now for the new function I added.

mmkekic reviewed (commented):
I think we should also have an exact test for t_evol dataframe. 
Also, in the case there are no events in a given time slice (ie datahang) the code will fail. We can either raise an exception or do some default behavior (ie set all t_evol params to 0s or omit the time slice all together) and print a warning, maybe the latter is more convenient for the actual map production.

bpalmeiro committed:
Make masks names consistent

bpalmeiro committed:
Cosmetics

bpalmeiro commented:
> I think we should also have an exact test for t_evol dataframe.

I agree, on that!

> Also, in the case there are no events in a given time slice (ie data hang) the code will fail. We can either raise an exception or do some default behavior (ie set all t_evol params to 0s or omit the time slice all together) and print a warning, maybe the latter is more convenient for the actual map production.

I think this belongs to another PR

bpalmeiro committed:
Add test comparing exact result for evolution tables

bpalmeiro commented:
A test comparing exact results is done. I lowered the time bit for the evolution table, so we get two raws in order to avoid possible issues with several entries, as saw during the writing of these functions and tests.

bpalmeiro committed:
Add tests: per cut and per bin.

The different_cuts tests the value in different cuts columns are within
expected, the different_time_bins tests the values in different time
bins are within expected

mmkekic committed:
Add test for get_time_series_df using pandas cut

bpalmeiro committed:
Make add_krevol work with uncut dst

kwargs are added  as an input, so the code is backward compatible in
terms of configuration files, they play no role inside the function.

mmkekic committed:
Add test for cut_time_evolution with multiple events.

The tests check that the result of efficiencies does not depend on
whether the events are repeated in the dataframe.

mmkekic commented on krcal/core/kr_parevol_functions_test.py:
```diff
@@ -0,0 +1,97 @@
+import pandas as pd
+
+from numpy.testing              import assert_allclose
+from flaky                      import flaky
```
Maybe we should use flaky in the first two tests

mmkekic commented on krcal/core/kr_parevol_functions_test.py:
```diff
@@ -0,0 +1,97 @@
+def test_cut_time_evolution_different_cuts(folder_test_dst,
+                                           test_dst_file ):
+
+    dst                = load_dst(folder_test_dst + test_dst_file, "DST", "Events")
```
since we use load_dst 3 times, probably better to put it in a fixture

mmkekic commented on krcal/core/kr_parevol_functions_test.py:
```diff
@@ -0,0 +1,124 @@
+
+    dst             = load_dst(folder_test_dst + test_dst_file, "DST", "Events")
+    mask_S1         = dst         .nS1   == 1
+    mask_S2         = dst[mask_S1].nS2   == 1
```
Iam a bit confused with the mask_S2. As it is defined now it will be pandas Series with smaller size than mask_S1. The reason why you could still do mask_S1 & mask_S2 is because pandas will keep False by default if the rows are not contained in both series, so technically is not wrong but is also not what we do in apply_cuts function. Also it would not work if mask_S1 and mask_S2 were numpy arrays. I dont think this tests brings anything that is not contained in the previous test, but if you think it is needed, fix the dimensions and aligning and im happy to approve this PR :)

bpalmeiro replied:
Yes, the mask was not intended to be a subset. I agree it does not add any check to the previous test, I will proceed to remove it.

bpalmeiro replied:
Should I remove and keep it in the commit line, or is it better to pretend it never existed?

mmkekic reviewed (approved):
This PR adds cut efficiencies information to t_evol, which is needed to be able to track efficiencies changes in time. The code is well documented and tested and I am happy to approve it. Good job!
